# Data Cleaning for Segmenting credit card customers using K-Means in Python - Guided Project

### 🧼 Side Project: Data Cleaning for Guided Project Alignment

While working on a guided project, I did not have access to the clean dataset referenced in the instructions. However, I was fortunate to discover a highly similar dataset on Kaggle:

**Dataset:** [Credit Card Customer Data – Kaggle](https://www.kaggle.com/datasets/sakshigoyal7/credit-card-customers?resource=download)  
**Downloaded on:** 26/05/2025

To proceed with the guided project, I completed a side project that focused on cleaning and preprocessing this dataset to closely mirror the structure and quality of the original reference data. This allowed me to effectively follow the project requirements while reinforcing essential data preparation skills.

In [1]:
import pandas as pd        # For data manipulation and analysis
import numpy as np         # For numerical operations, especially with arrays and statistical functions
import matplotlib.pyplot as plt  # For data visualization and plotting

In [2]:
# Load the dataset into a DataFrame
BankChurners_df = pd.read_csv('BankChurners.csv')

# Display the first five rows to get an initial look at the data structure
BankChurners_df.head()

CLIENTNUM     Attrition_Flag  Customer_Age Gender  Dependent_count  \
0  768805383  Existing Customer            45      M                3   
1  818770008  Existing Customer            49      F                5   
2  713982108  Existing Customer            51      M                3   
3  769911858  Existing Customer            40      F                4   
4  709106358  Existing Customer            40      M                3   

  Education_Level Marital_Status Income_Category Card_Category  \
0     High School        Married     $60K - $80K          Blue   
1        Graduate         Single  Less than $40K          Blue   
2        Graduate        Married    $80K - $120K          Blue   
3     High School        Unknown  Less than $40K          Blue   
4      Uneducated        Married     $60K - $80K          Blue   

   Months_on_book  ...  Credit_Limit  Total_Revolving_Bal  Avg_Open_To_Buy  \
0              39  ...       12691.0                  777          11914.0   
1              44  ...        8256.0                  864           7392.0   
2              36  ...        3418.0                    0           3418.0   
3              34  ...        3313.0                 2517            796.0   
4              21  ...        4716.0                    0           4716.0   

   Total_Amt_Chng_Q4_Q1  Total_Trans_Amt  Total_Trans_Ct  Total_Ct_Chng_Q4_Q1  \
0                 1.335             1144              42                1.625   
1                 1.541             1291              33                3.714   
2                 2.594             1887              20                2.333   
3                 1.405             1171              20                2.333   
4                 2.175              816              28                2.500   

   Avg_Utilization_Ratio  \
0                  0.061   
1                  0.105   
2                  0.000   
3                  0.760   
4                  0.000   

   Naive_Bayes_Classifier_Attrition_Flag_Card_Category_Contacts_Count_12_mon_Dependent_count_Education_Level_Months_Inactive_12_mon_1  \
0                                           0.000093                                                                                    
1                                           0.000057                                                                                    
2                                           0.000021                                                                                    
3                                           0.000134                                                                                    
4                                           0.000022                                                                                    

   Naive_Bayes_Classifier_Attrition_Flag_Card_Category_Contacts_Count_12_mon_Dependent_count_Education_Level_Months_Inactive_12_mon_2  
0                                            0.99991                                                                                   
1                                            0.99994                                                                                   
2                                            0.99998                                                                                   
3                                            0.99987                                                                                   
4                                            0.99998                                                                                   

[5 rows x 23 columns]

In [3]:
# Get a concise summary of the dataset
# Includes number of entries, column names, non-null counts, and data types
BankChurners_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10127 entries, 0 to 10126
Data columns (total 23 columns):
 #   Column                                                                                                                              Non-Null Count  Dtype  
---  ------                                                                                                                              --------------  -----  
 0   CLIENTNUM                                                                                                                           10127 non-null  int64  
 1   Attrition_Flag                                                                                                                      10127 non-null  object 
 2   Customer_Age                                                                                                                        10127 non-null  int64  
 3   Gender                                                                           

In [4]:
# Select relevant features from the original dataset for analysis
# These features align with the structure and variables used in the Dataquest guided project
credit_card_df = BankChurners_df[[
    'CLIENTNUM',                  # Unique customer identifier
    'Customer_Age',               # Age of the customer
    'Gender',                     # Gender of the customer
    'Dependent_count',            # Number of dependents
    'Education_Level',            # Education background (useful for income or churn patterns)
    'Marital_Status',             # Relationship status
    'Income_Category',            # Customer's income bracket
    'Months_on_book',             # Tenure with the bank
    'Total_Relationship_Count',   # Total number of bank products/services used
    'Months_Inactive_12_mon',     # Inactive months in the past year
    'Contacts_Count_12_mon',      # Number of customer service contacts
    'Credit_Limit',               # Credit limit on the account
    'Total_Trans_Amt',            # Total transaction amount over a defined period
    'Total_Trans_Ct',             # Total number of transactions
    'Avg_Utilization_Ratio'       # Average credit utilization ratio
]]

In [5]:
# Preview the first five rows of the selected feature set
# This helps confirm the selection was successful and provides a quick sanity check on values and formats
credit_card_df.head()

CLIENTNUM  Customer_Age Gender  Dependent_count Education_Level  \
0  768805383            45      M                3     High School   
1  818770008            49      F                5        Graduate   
2  713982108            51      M                3        Graduate   
3  769911858            40      F                4     High School   
4  709106358            40      M                3      Uneducated   

  Marital_Status Income_Category  Months_on_book  Total_Relationship_Count  \
0        Married     $60K - $80K              39                         5   
1         Single  Less than $40K              44                         6   
2        Married    $80K - $120K              36                         4   
3        Unknown  Less than $40K              34                         3   
4        Married     $60K - $80K              21                         5   

   Months_Inactive_12_mon  Contacts_Count_12_mon  Credit_Limit  \
0                       1                      3       12691.0   
1                       1                      2        8256.0   
2                       1                      0        3418.0   
3                       4                      1        3313.0   
4                       1                      0        4716.0   

   Total_Trans_Amt  Total_Trans_Ct  Avg_Utilization_Ratio  
0             1144              42                  0.061  
1             1291              33                  0.105  
2             1887              20                  0.000  
3             1171              20                  0.760  
4              816              28                  0.000

In [6]:
# Display a concise summary of the selected features DataFrame
# Includes number of entries, column data types, and non-null value counts
# Useful for spotting missing values and confirming data types before preprocessing
credit_card_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10127 entries, 0 to 10126
Data columns (total 15 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   CLIENTNUM                 10127 non-null  int64  
 1   Customer_Age              10127 non-null  int64  
 2   Gender                    10127 non-null  object 
 3   Dependent_count           10127 non-null  int64  
 4   Education_Level           10127 non-null  object 
 5   Marital_Status            10127 non-null  object 
 6   Income_Category           10127 non-null  object 
 7   Months_on_book            10127 non-null  int64  
 8   Total_Relationship_Count  10127 non-null  int64  
 9   Months_Inactive_12_mon    10127 non-null  int64  
 10  Contacts_Count_12_mon     10127 non-null  int64  
 11  Credit_Limit              10127 non-null  float64
 12  Total_Trans_Amt           10127 non-null  int64  
 13  Total_Trans_Ct            10127 non-null  int64  
 14  Avg_Ut

### 🔍 Initial Data Inspection Observations

Based on the `.info()` output for the selected dataset, there are no missing values in the current feature set. However, from experience, columns with an `object` data type may still contain inconsistencies or unexpected entries—such as typos, placeholder strings (e.g., "Unknown" or "N/A"), or formatting issues. These should be carefully reviewed and cleaned before encoding or analysis to ensure data quality.

In [7]:
# Check the distribution of values in the 'Gender' column
# Useful for spotting unexpected entries or typos (e.g., inconsistent casing or missing categories)
credit_card_df['Gender'].value_counts()

Gender
F    5358
M    4769
Name: count, dtype: int64

In [8]:
# Examine the distribution of values in the 'Education_Level' column
# Useful for identifying categories, spotting inconsistencies (e.g. "Unknown", "Uneducated", etc.), and planning encoding or grouping strategies
credit_card_df['Education_Level'].value_counts()

Education_Level
Graduate         3128
High School      2013
Unknown          1519
Uneducated       1487
College          1013
Post-Graduate     516
Doctorate         451
Name: count, dtype: int64

### 📊 Comparison of 'Education_Level' Between My Dataset and the Guided Project Dataset

Upon comparison, I observed that the **guided project dataset contains no 'Unknown' values** in the `Education_Level` column, whereas my dataset includes them. This discrepancy affects the counts of each known education category.

| Education Level | My Data | Dataquest Data | Difference |
|-----------------|---------|----------------|------------|
| Graduate        | 3,128   | 3,685          | -557       |
| High School     | 2,013   | 2,351          | -338       |
| Uneducated      | 1,487   | 1,755          | -268       |
| College         | 1,013   | 1,192          | -179       |
| Post Graduate   |   516   |   616          | -100       |
| Doctorate       |   451   |   528          |  -77       |

**Total discrepancy across categories:** 1,519  
This total precisely matches the number of `'Unknown'` values present in my dataset, indicating that those entries were either removed or reclassified in the guided project data.

In [9]:
# Analyze the distribution of values in the 'Marital_Status' column
# Helps identify all possible categories, detect inconsistencies (e.g., 'Unknown' or typos), and guide encoding strategies
credit_card_df['Marital_Status'].value_counts()

Marital_Status
Married     4687
Single      3943
Unknown      749
Divorced     748
Name: count, dtype: int64

In [10]:
# Examine the distribution of values in the 'Income_Category' column
# This helps uncover income brackets present in the data and spot any anomalies or placeholders like 'Unknown'
credit_card_df['Income_Category'].value_counts()

Income_Category
Less than $40K    3561
$40K - $60K       1790
$80K - $120K      1535
$60K - $80K       1402
Unknown           1112
$120K +            727
Name: count, dtype: int64

### 💡 Observation on Income Data Discrepancy

In comparing my dataset to the one used in the guided project, I observed a key structural difference:

- My dataset contains an `Income_Category` column, which includes bracketed income levels and some `'Unknown'` values.
- The Dataquest dataset, however, provides an `Estimated_Income` column with **continuous income values** and **no missing entries**.

This distinction affects how income data can be used in analysis. While the bracketed categories require encoding or grouping, the estimated values in the guided project dataset offer more granularity for modeling or segmentation.

## Data Cleaning process

### Cleaning the Education Level

In [11]:
# Analyze records where Education_Level is known,  
# grouping by Income_Category and Gender to examine distribution.  
# Count occurrences, reset index for clarity, and sort by Income_Category.  

(credit_card_df[credit_card_df['Education_Level']!='Unknown'][['Income_Category', 'Gender']]
 .value_counts().reset_index().sort_values(by=['Income_Category']))

Income_Category Gender  count
6         $120K +      M    607
4     $40K - $60K      F    859
5     $40K - $60K      M    662
2     $60K - $80K      M   1192
1    $80K - $120K      M   1316
0  Less than $40K      F   2767
7  Less than $40K      M    238
3         Unknown      F    920
8         Unknown      M     47

In [12]:
# Analyze records where Education_Level is 'Unknown',  
# grouping by Income_Category and Gender to understand distribution.  
# Count occurrences, reset index for readability, and sort by Income_Category.  

(credit_card_df[credit_card_df['Education_Level']=='Unknown'][['Income_Category', 'Gender']]
 .value_counts().reset_index().sort_values(by=['Income_Category']))

Income_Category Gender  count
5         $120K +      M    120
3     $40K - $60K      F    155
6     $40K - $60K      M    114
2     $60K - $80K      M    210
1    $80K - $120K      M    219
0  Less than $40K      F    517
7  Less than $40K      M     39
4         Unknown      F    140
8         Unknown      M      5

### 🎯 Insight on Handling Unknowns in Education Level

After analyzing the dataset and reviewing the comparative tables, it became clear that **income category** and **gender** are closely linked with a customer's education level. This insight guided my approach to handling entries labeled as `'Unknown'` in the `education_level` column.

By leveraging patterns observed across these related features, I was able to develop a more informed imputation strategy that redistributed unknown education entries into likely categories—while maintaining the overall data integrity and demographic balance.

In [13]:
# Estimating education level distribution for unknown cases based on Income Category and Gender  

# Copy the original dataframe to preserve raw data  
credit_clean = credit_card_df.copy()

# Define income categories and gender groups for iteration  
Income_Cat = ['$120K +', '$80K - $120K', '$60K - $80K', '$40K - $60K', 'Less than $40K', 'Unknown'] 
Genders = ['M', 'F'] 

# List to store processed customer data  
all_customer_data = []

# Iterate over each gender 
for Gender in Genders: # Outer loop for Gender
    for Income in Income_Cat: # Inner loop for Income Category
        # Filter for customers with known education levels within the current gender and income category
        filtered_data = credit_clean[
            (credit_clean['Income_Category'] == Income) &
            (credit_clean['Education_Level'] != 'Unknown') &
            (credit_clean['Gender'] == Gender) # Use the Gender from the outer loop
        ]

        # Count total customers in the filtered subset 
        total_customer = filtered_data.shape[0]

        # Proceed only if there are customers in the filtered subset
        if total_customer > 0: 
            # Count occurrences of each education level
            
            data = filtered_data['Education_Level'].value_counts().reset_index()
            data.columns = ['Education_Level', 'count']

            # Assign corresponding income category and gender
            data['Income_Category'] = Income
            data['Gender'] = Gender 

            # Compute client ratio based on total customers in the category
            data['client_ratio'] = data['count'] / total_customer
            
            # Store processed data 
            all_customer_data.append(data)
        else:
            # Inform user when no matching customers exist
            print(f"No {Gender} customer with known education level found for Income Category: {Income}")

# Combine all collected data into a single dataframe 
customer = pd.concat(all_customer_data, ignore_index=True)

# Display the final dataset 
customer

No F customer with known education level found for Income Category: $120K +
No F customer with known education level found for Income Category: $80K - $120K
No F customer with known education level found for Income Category: $60K - $80K


Education_Level  count Income_Category Gender  client_ratio
0         Graduate    204         $120K +      M      0.336079
1      High School    147         $120K +      M      0.242175
2       Uneducated    119         $120K +      M      0.196046
3          College     70         $120K +      M      0.115321
4        Doctorate     37         $120K +      M      0.060956
5    Post-Graduate     30         $120K +      M      0.049423
6         Graduate    478    $80K - $120K      M      0.363222
7      High School    308    $80K - $120K      M      0.234043
8       Uneducated    217    $80K - $120K      M      0.164894
9          College    175    $80K - $120K      M      0.132979
10   Post-Graduate     81    $80K - $120K      M      0.061550
11       Doctorate     57    $80K - $120K      M      0.043313
12        Graduate    422     $60K - $80K      M      0.354027
13     High School    307     $60K - $80K      M      0.257550
14      Uneducated    195     $60K - $80K      M      0.163591
15         College    132     $60K - $80K      M      0.110738
16   Post-Graduate     77     $60K - $80K      M      0.064597
17       Doctorate     59     $60K - $80K      M      0.049497
18        Graduate    234     $40K - $60K      M      0.353474
19     High School    160     $40K - $60K      M      0.241692
20      Uneducated    116     $40K - $60K      M      0.175227
21         College     75     $40K - $60K      M      0.113293
22   Post-Graduate     50     $40K - $60K      M      0.075529
23       Doctorate     27     $40K - $60K      M      0.040785
24        Graduate    100  Less than $40K      M      0.420168
25     High School     54  Less than $40K      M      0.226891
26      Uneducated     37  Less than $40K      M      0.155462
27         College     26  Less than $40K      M      0.109244
28       Doctorate     11  Less than $40K      M      0.046218
29   Post-Graduate     10  Less than $40K      M      0.042017
30        Graduate     20         Unknown      M      0.425532
31     High School      9         Unknown      M      0.191489
32      Uneducated      7         Unknown      M      0.148936
33   Post-Graduate      5         Unknown      M      0.106383
34         College      3         Unknown      M      0.063830
35       Doctorate      3         Unknown      M      0.063830
36        Graduate    319     $40K - $60K      F      0.371362
37     High School    195     $40K - $60K      F      0.227008
38      Uneducated    133     $40K - $60K      F      0.154831
39         College    108     $40K - $60K      F      0.125728
40   Post-Graduate     61     $40K - $60K      F      0.071013
41       Doctorate     43     $40K - $60K      F      0.050058
42        Graduate   1039  Less than $40K      F      0.375497
43     High School    617  Less than $40K      F      0.222985
44      Uneducated    485  Less than $40K      F      0.175280
45         College    319  Less than $40K      F      0.115287
46   Post-Graduate    160  Less than $40K      F      0.057824
47       Doctorate    147  Less than $40K      F      0.053126
48        Graduate    312         Unknown      F      0.339130
49     High School    216         Unknown      F      0.234783
50      Uneducated    178         Unknown      F      0.193478
51         College    105         Unknown      F      0.114130
52       Doctorate     67         Unknown      F      0.072826
53   Post-Graduate     42         Unknown      F      0.045652

In [14]:
# Convert 'Unknown' values to NaN to facilitate missing value handling and iteration
credit_clean['Education_Level'].replace('Unknown', np.NaN, inplace = True)
credit_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10127 entries, 0 to 10126
Data columns (total 15 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   CLIENTNUM                 10127 non-null  int64  
 1   Customer_Age              10127 non-null  int64  
 2   Gender                    10127 non-null  object 
 3   Dependent_count           10127 non-null  int64  
 4   Education_Level           8608 non-null   object 
 5   Marital_Status            10127 non-null  object 
 6   Income_Category           10127 non-null  object 
 7   Months_on_book            10127 non-null  int64  
 8   Total_Relationship_Count  10127 non-null  int64  
 9   Months_Inactive_12_mon    10127 non-null  int64  
 10  Contacts_Count_12_mon     10127 non-null  int64  
 11  Credit_Limit              10127 non-null  float64
 12  Total_Trans_Amt           10127 non-null  int64  
 13  Total_Trans_Ct            10127 non-null  int64  
 14  Avg_Ut

In [15]:
# Imputing missing Education_Level values using probabilities derived from known data  

# Define income categories and gender groups for iteration
Income_Cat = ['$120K +', '$80K - $120K', '$60K - $80K', '$40K - $60K', 'Less than $40K', 'Unknown']
Genders = ['M', 'F']

# Iterate through gender and income category combinations 
for gender in Genders:
    for income in Income_Cat:
        # Create a mask to filter records with missing education level for the current group
        missing_education_for_group_mask = (
            (credit_clean['Income_Category'] == income) &
            (credit_clean['Gender'] == gender) &
            (credit_clean['Education_Level'].isnull())
        )

        # Count how many records are missing Education_Level for this group
        num_missing_education = missing_education_for_group_mask.sum()

        # Proceed only if there are missing values to impute
        if num_missing_education > 0:
            
            # Retrieve known education distributions for the current group
            filter_probabilities = customer[
                (customer['Income_Category'] == income) &
                (customer['Gender'] == gender)
            ]
            
            # Extract possible education levels and their corresponding probabilities  
            education_choices = filter_probabilities['Education_Level'].tolist()
            education_probabilities = filter_probabilities['client_ratio'].tolist()

            # Ensure valid choices exist and probabilities are meaningful
            if education_choices and sum(education_probabilities) > 0: # Ensure probabilities are not all zero
                # Normalise probabilities to sum to 1 if needed (avoiding floating-point inconsistencies)
                prob_sum = sum(education_probabilities)
                if prob_sum != 1.0:
                    education_probabilities = [p / prob_sum for p in education_probabilities]

                # Randomly assign Education_Level values based on observed probabilities
                random_imputations = np.random.choice(
                    a=education_choices,
                    p=education_probabilities,
                    size=num_missing_education
                )

                # Assign imputed values directly using the mask 
                credit_clean.loc[missing_education_for_group_mask, 'Education_Level'] = random_imputations

                print(f"\nSuccessfully imputed {num_missing_education} for income range '{income}' and gender '{gender}' missing 'Education_Level' values.")
            else:
                print(f"\nSkipping imputation for income range '{income}' and gender '{gender}': No valid education choices or probabilities found in 'client' DataFrame.")
        else:
            print(f"\nNo missing 'Education_Level' for income range '{income}' and gender '{gender}' values found to impute.")

# Display updated dataframe structure after imputations 
print("\n--- data_test_imputed after imputation ---")
credit_clean.info()


Successfully imputed 120 for income range '$120K +' and gender 'M' missing 'Education_Level' values.

Successfully imputed 219 for income range '$80K - $120K' and gender 'M' missing 'Education_Level' values.

Successfully imputed 210 for income range '$60K - $80K' and gender 'M' missing 'Education_Level' values.

Successfully imputed 114 for income range '$40K - $60K' and gender 'M' missing 'Education_Level' values.

Successfully imputed 39 for income range 'Less than $40K' and gender 'M' missing 'Education_Level' values.

Successfully imputed 5 for income range 'Unknown' and gender 'M' missing 'Education_Level' values.

No missing 'Education_Level' for income range '$120K +' and gender 'F' values found to impute.

No missing 'Education_Level' for income range '$80K - $120K' and gender 'F' values found to impute.

No missing 'Education_Level' for income range '$60K - $80K' and gender 'F' values found to impute.

Successfully imputed 155 for income range '$40K - $60K' and gender 'F' mi

In [16]:
# Review the updated distribution of Education_Level after imputation to verify replacement of missing values  
credit_clean['Education_Level'].value_counts()

Education_Level
Graduate         3702
High School      2359
Uneducated       1760
College          1183
Post-Graduate     605
Doctorate         518
Name: count, dtype: int64

### 📊 Comparison Between My Clean Dataset and the Guided Project Dataset

Below is a comparison of `Education_Level` counts between my cleaned dataset and the Dataquest dataset:

| Education Level | My Data | Dataquest Data | Difference |
|-----------------|---------|----------------|------------|
| Graduate        | 3,687   | 3,685          | 2          |
| High School     | 2,349   | 2,351          | -2         |
| Uneducated      | 1,756   | 1,755          | 1          |
| College         | 1,192   | 1,192          | 0          |
| Post Graduate   |   617   |   616          | 1          |
| Doctorate       |   526   |   528          | -2         |

**Observation:** While individual category counts vary slightly with each run—due to the randomness introduced during imputation—the **total distribution remains balanced**, with the differences summing to zero. This confirms that the imputation approach preserves dataset size while subtly shifting the category proportions.

### Cleaning the Income Category

In [17]:
# Remove dollar signs and 'K' from Income_Category values to standardise format for analysis
minus_kd=credit_clean['Income_Category'].apply(
    lambda x: x.replace('K', '').replace('$', '')
)
minus_kd_df = pd.DataFrame(minus_kd)

In [18]:
# Map cleaned income values to descriptive income level categories for clearer interpretation
Income_mapping = {'Less than 40' : 'Survival Income',
                  '40 - 60' : 'Basic Needs Income',
                  '60 - 80' : 'Discretionary Income',
                  '80 - 120' : 'Significant Income',
                  '120 +' : 'Abundance Income',
                  'Unknown' : 'Unknown'
                 }
minus_kd_df['Income_level'] = minus_kd_df['Income_Category'].map(Income_mapping)

In [19]:
# Preview the first five rows to verify cleaned and mapped income levels
minus_kd_df.head()

Income_Category          Income_level
0         60 - 80  Discretionary Income
1    Less than 40       Survival Income
2        80 - 120    Significant Income
3    Less than 40       Survival Income
4         60 - 80  Discretionary Income

In [20]:
# Insert the mapped Income_level column into credit_clean at position 7 for consistent column ordering
credit_clean.insert(7, 'Income_level', minus_kd_df['Income_level'])

In [21]:
# Count occurrences of each Income_level category and reset the index for a tidy DataFrame view
credit_clean['Income_level'].value_counts().reset_index()

Income_level  count
0       Survival Income   3561
1    Basic Needs Income   1790
2    Significant Income   1535
3  Discretionary Income   1402
4               Unknown   1112
5      Abundance Income    727

In [22]:
# Remove the Income_Category column as it is no longer needed after mapping to Income_level
credit_clean = credit_clean.drop("Income_Category", axis=1)

In [23]:
# Explore distribution of Education_Level and Gender for customers with unknown Income_level,  
# count combinations, tidy the output, and sort by Education_Level for easy interpretation

(credit_clean[credit_clean['Income_level'] == 'Unknown' ]
 [['Education_Level','Gender']].value_counts().reset_index().sort_values(by=['Education_Level']))

Education_Level Gender  count
3          College      F    122
10         College      M      4
4        Doctorate      F     73
11       Doctorate      M      3
0         Graduate      F    361
6         Graduate      M     20
1      High School      F    244
7      High School      M     12
5    Post-Graduate      F     52
9    Post-Graduate      M      6
2       Uneducated      F    208
8       Uneducated      M      7

In [24]:
# Replace 'Unknown' in Income_level with NaN to treat it as missing data for further analysis or imputation
credit_clean['Income_level'].replace('Unknown', np.NaN, inplace = True)
credit_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10127 entries, 0 to 10126
Data columns (total 15 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   CLIENTNUM                 10127 non-null  int64  
 1   Customer_Age              10127 non-null  int64  
 2   Gender                    10127 non-null  object 
 3   Dependent_count           10127 non-null  int64  
 4   Education_Level           10127 non-null  object 
 5   Marital_Status            10127 non-null  object 
 6   Income_level              9015 non-null   object 
 7   Months_on_book            10127 non-null  int64  
 8   Total_Relationship_Count  10127 non-null  int64  
 9   Months_Inactive_12_mon    10127 non-null  int64  
 10  Contacts_Count_12_mon     10127 non-null  int64  
 11  Credit_Limit              10127 non-null  float64
 12  Total_Trans_Amt           10127 non-null  int64  
 13  Total_Trans_Ct            10127 non-null  int64  
 14  Avg_Ut

In [25]:
# Calculate income level distribution by Education_Level and Gender to guide imputation for unknown income entries

# Define education levels and gender groups for iteration 
Education_Level = ['College', 'Doctorate', 'Graduate', 'High School', 'Post-Graduate', 'Uneducated']
Genders = ['M', 'F']

# List to collect processed income distribution data  
all_customer_data = []

# Iterate over each gender group  
for Gender in Genders: # Outer loop for Gender
    for ed_level in Education_Level: # Inner loop for Income Category
        # Filter for records with the current education level and gender, excluding unknown income levels  
        filtered_data = credit_clean[
            (credit_clean['Education_Level'] == ed_level) &
            (credit_clean['Income_level'] != 'Unknown') &
            (credit_clean['Gender'] == Gender) # Use the Gender from the outer loop
        ]

        # Count total customers matching current filter criteria  
        total_customer = filtered_data.shape[0]
        
        if total_customer > 0:
            # Count frequency of each income level in the filtered subset
            data = filtered_data['Income_level'].value_counts().reset_index()
            data.columns = ['Income_level', 'count']

            # Append contextual attributes
            data['Education_Level'] = ed_level
            data['Gender'] = Gender 

            # Calculate the proportion of customers in each income level
            data['client_ratio'] = data['count'] / total_customer

            all_customer_data.append(data)
        else:
            # Notify if no qualifying customers are found for the current group
            print(f"No {Gender} customers with known education level found for Income Category: {Income}")

# Combine all the income distribution data into a single DataFrame
customer = pd.concat(all_customer_data, ignore_index=True)

# Display the final grouped and weighted income distribution
customer

Income_level  count Education_Level Gender  client_ratio
0     Significant Income    197         College      M      0.352415
1   Discretionary Income    153         College      M      0.273703
2     Basic Needs Income     91         College      M      0.162791
3       Abundance Income     85         College      M      0.152057
4        Survival Income     29         College      M      0.051878
5   Discretionary Income     68       Doctorate      M      0.296943
6     Significant Income     63       Doctorate      M      0.275109
7       Abundance Income     48       Doctorate      M      0.209607
8     Basic Needs Income     34       Doctorate      M      0.148472
9        Survival Income     13       Doctorate      M      0.056769
10    Significant Income    564        Graduate      M      0.327907
11  Discretionary Income    505        Graduate      M      0.293605
12    Basic Needs Income    278        Graduate      M      0.161628
13      Abundance Income    239        Graduate      M      0.138953
14       Survival Income    114        Graduate      M      0.066279
15  Discretionary Income    360     High School      M      0.310078
16    Significant Income    357     High School      M      0.307494
17    Basic Needs Income    187     High School      M      0.161068
18      Abundance Income    181     High School      M      0.155900
19       Survival Income     64     High School      M      0.055125
20    Significant Income     98   Post-Graduate      M      0.336770
21  Discretionary Income     86   Post-Graduate      M      0.295533
22    Basic Needs Income     56   Post-Graduate      M      0.192440
23      Abundance Income     34   Post-Graduate      M      0.116838
24       Survival Income     11   Post-Graduate      M      0.037801
25    Significant Income    256      Uneducated      M      0.316440
26  Discretionary Income    230      Uneducated      M      0.284302
27      Abundance Income    140      Uneducated      M      0.173053
28    Basic Needs Income    130      Uneducated      M      0.160692
29       Survival Income     46      Uneducated      M      0.056860
30       Survival Income    376         College      F      0.602564
31    Basic Needs Income    126         College      F      0.201923
32       Survival Income    166       Doctorate      F      0.574394
33    Basic Needs Income     50       Doctorate      F      0.173010
34       Survival Income   1238        Graduate      F      0.624622
35    Basic Needs Income    383        Graduate      F      0.193239
36       Survival Income    717     High School      F      0.598497
37    Basic Needs Income    237     High School      F      0.197830
38       Survival Income    192   Post-Graduate      F      0.611465
39    Basic Needs Income     70   Post-Graduate      F      0.222930
40       Survival Income    595      Uneducated      F      0.625657
41    Basic Needs Income    148      Uneducated      F      0.155626

In [26]:
# Impute missing Income_level values based on known distributions by Education_Level and Gender

# Define education levels and gender groups for iteration 
Education_Level = ['College', 'Doctorate', 'Graduate', 'High School', 'Post-Graduate', 'Uneducated']
Genders = ['M', 'F']

# Iterate through gender and education levels combinations 
for gender in Genders:
    for ed_level in Education_Level:
        # Identify records with missing income level for the current education level and gender  
        missing_income_for_group_mask = (
            (credit_clean['Education_Level'] == ed_level) &
            (credit_clean['Gender'] == gender) &
            (credit_clean['Income_level'].isnull())
        )

        # Total missing entries to impute 
        num_missing_income = missing_income_for_group_mask.sum()

        # Proceed only if there are missing income records to fill
        if num_missing_income > 0:
            # Retrieve the income distribution for this group from the 'customer' DataFrame  
            filter_probabilities = customer[
                (customer['Education_Level'] == ed_level) &
                (customer['Gender'] == gender)
            ]

            income_choices = filter_probabilities['Income_level'].tolist()
            income_probabilities = filter_probabilities['client_ratio'].tolist()
           
            # Ensure we have valid income categories and non-zero probabilities
            if income_choices and sum(income_probabilities) > 0: # Ensure probabilities are not all zero
                # Normalise probabilities to ensure they sum to 1
                prob_sum = sum(income_probabilities)
                if prob_sum != 1.0:
                    income_probabilities = [p / prob_sum for p in income_probabilities]

                # Generate random imputed values using observed income distribution
                random_imputations = np.random.choice(
                    a=income_choices,
                    p=income_probabilities,
                    size=num_missing_income
                )

                # Assign the imputed values back to the appropriate rows in credit_clean  
                credit_clean.loc[missing_income_for_group_mask, 'Income_level'] = random_imputations

                print(f"\nSuccessfully imputed {num_missing_income} for education level '{ed_level}' and gender '{gender}' missing 'Income_Level' values.")
            else:
                print(f"\nSkipping imputation for education level '{ed_level}' and gender '{gender}': No valid education choices or probabilities found in 'client' DataFrame.")
        else:
            print(f"\nNo missing 'Income_level' for education level '{ed_level}' and gender '{gender}' values found to impute.")

# Display the structure of the DataFrame after imputation
print("\n--- data_test_imputed after imputation ---")
credit_clean.info()



Successfully imputed 4 for education level 'College' and gender 'M' missing 'Income_Level' values.

Successfully imputed 3 for education level 'Doctorate' and gender 'M' missing 'Income_Level' values.

Successfully imputed 20 for education level 'Graduate' and gender 'M' missing 'Income_Level' values.

Successfully imputed 12 for education level 'High School' and gender 'M' missing 'Income_Level' values.

Successfully imputed 6 for education level 'Post-Graduate' and gender 'M' missing 'Income_Level' values.

Successfully imputed 7 for education level 'Uneducated' and gender 'M' missing 'Income_Level' values.

Successfully imputed 122 for education level 'College' and gender 'F' missing 'Income_Level' values.

Successfully imputed 73 for education level 'Doctorate' and gender 'F' missing 'Income_Level' values.

Successfully imputed 361 for education level 'Graduate' and gender 'F' missing 'Income_Level' values.

Successfully imputed 244 for education level 'High School' and gender 'F'

In [27]:
# Count occurrences of each Income_level category and reset the index for a well-structured DataFrame  
credit_clean['Income_level'].value_counts().reset_index()

Income_level  count
0       Survival Income   4376
1    Basic Needs Income   2047
2    Significant Income   1556
3  Discretionary Income   1414
4      Abundance Income    734

### Changing Columns name to matches with dataquest data

In [28]:
# Rename dataset columns to follow Dataquest naming conventions for consistency and clarity
# 'income_level' is a custom feature derived from income categories, serving as a proxy for estimated income
credit_clean.rename(columns={
    'CLIENTNUM': 'customer_id',                      # Unique identifier for each customer
    'Customer_Age': 'age',                           # Renamed for clarity and consistency
    'Gender': 'gender',
    'Dependent_count': 'dependent_count',
    'Education_Level': 'education_level',
    'Marital_Status': 'marital_status',
    'Income_level': 'income_level',                  # Custom classification based on income category
    'Months_on_book': 'months_on_book',
    'Total_Relationship_Count': 'total_relationship_count',
    'Months_Inactive_12_mon': 'months_inactive_12_mon',
    'Contacts_Count_12_mon': 'contacts_count_12_mon',
    'Credit_Limit': 'credit_limit',
    'Total_Trans_Amt': 'total_trans_amount',
    'Total_Trans_Ct': 'total_trans_count',
    'Avg_Utilization_Ratio': 'avg_utilization_ratio'
}, inplace=True)

# Preview the updated DataFrame to confirm successful renaming
credit_clean.head()

customer_id  age gender  dependent_count education_level marital_status  \
0    768805383   45      M                3     High School        Married   
1    818770008   49      F                5        Graduate         Single   
2    713982108   51      M                3        Graduate        Married   
3    769911858   40      F                4     High School        Unknown   
4    709106358   40      M                3      Uneducated        Married   

           income_level  months_on_book  total_relationship_count  \
0  Discretionary Income              39                         5   
1       Survival Income              44                         6   
2    Significant Income              36                         4   
3       Survival Income              34                         3   
4  Discretionary Income              21                         5   

   months_inactive_12_mon  contacts_count_12_mon  credit_limit  \
0                       1                      3       12691.0   
1                       1                      2        8256.0   
2                       1                      0        3418.0   
3                       4                      1        3313.0   
4                       1                      0        4716.0   

   total_trans_amount  total_trans_count  avg_utilization_ratio  
0                1144                 42                  0.061  
1                1291                 33                  0.105  
2                1887                 20                  0.000  
3                1171                 20                  0.760  
4                 816                 28                  0.000

In [29]:
# Export the cleaned dataset to a CSV file named 'customer_segmentation.csv',  
# ensuring consistency with Dataquest file naming conventions  
credit_clean.to_csv('customer_segmentation.csv', index = False)

### 🧼 Summary of Dataset Preparation

To align with the dataset used in the Dataquest guided project, I sourced a similar dataset from Kaggle and applied a series of cleaning and transformation steps:

- Renamed columns to match the naming conventions of the Dataquest dataset for consistency.
- Successfully replicated the distribution of `education_level` categories using probabilistic imputation techniques.
- Unable to reproduce the `Estimated_Income` field, as my dataset only included categorical income brackets. As a workaround, I generated an `income_level` classification to approximate economic segmentation.
- Verified that both `gender` and `marital_status` distributions in my dataset exactly match those in the Dataquest dataset.